File: testplugins/game-plugin-demo/conformance/go/matrices/heartbeat_test.go
package matrices

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"

	"github.com/u-ai/mock-conformance-go"
)

func TestHeartbeat_Suite(t *testing.T) {
	cases := []conformance.TestCase{
		{
			ID:       "HB-001",
			Category: conformance.CatHeartbeat,
			Name:     "pause heartbeat for 500ms returns paused true and correct duration",
			Run: func(ctx context.Context, host *conformance.PseudoHost) (conformance.TestResult, error) {
				tr := conformance.TestResult{Category: string(conformance.CatHeartbeat)}
				result, err := heartbeatPauseRPC(ctx, 500)
				if err != nil {
					return tr, err
				}
				if result["paused"] != true {
					return tr, fmt.Errorf("expected paused true, got %v", result["paused"])
				}
				pauseMs, ok := result["pauseMs"].(float64)
				if !ok || pauseMs != 500 {
					return tr, fmt.Errorf("expected pauseMs 500, got %v", result["pauseMs"])
				}
				duration, ok := result["duration"].(string)
				if !ok || duration != "500ms" {
					return tr, fmt.Errorf("expected duration 500ms, got %v", result["duration"])
				}
				tr.Passed = true
				return tr, nil
			},
		},
		{
			ID:       "HB-002",
			Category: conformance.CatHeartbeat,
			Name:     "resume heartbeat after pause returns resumed true and wasPaused true",
			Run: func(ctx context.Context, host *conformance.PseudoHost) (conformance.TestResult, error) {
				tr := conformance.TestResult{Category: string(conformance.CatHeartbeat)}
				binary := conformance.FindPluginBinary()
				if binary == "" {
					return tr, fmt.Errorf("plugin binary not found")
				}
				h := conformance.NewPseudoHostForBinary(binary)
				if err := h.Start(ctx); err != nil {
					return tr, fmt.Errorf("start host: %w", err)
				}
				defer h.Kill()

				pausePayload := conformance.MustMarshal(map[string]any{"pauseMs": 2000})
				_, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.pause", pausePayload)
				if err != nil {
					return tr, fmt.Errorf("heartbeat.pause rpc: %w", err)
				}

				resumeResp, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.resume", conformance.MustMarshal(map[string]any{}))
				if err != nil {
					return tr, fmt.Errorf("heartbeat.resume rpc: %w", err)
				}
				var resumeResult map[string]any
				if err := json.Unmarshal(resumeResp.Payload, &resumeResult); err != nil {
					return tr, fmt.Errorf("unmarshal resume response: %w", err)
				}
				if resumeResult["resumed"] != true {
					return tr, fmt.Errorf("expected resumed true, got %v", resumeResult["resumed"])
				}
				if resumeResult["wasPaused"] != true {
					return tr, fmt.Errorf("expected wasPaused true, got %v", resumeResult["wasPaused"])
				}
				tr.Passed = true
				return tr, nil
			},
		},
		{
			ID:       "HB-003",
			Category: conformance.CatHeartbeat,
			Name:     "drop heartbeat with count 5 returns configured true and correct dropCount",
			Run: func(ctx context.Context, host *conformance.PseudoHost) (conformance.TestResult, error) {
				tr := conformance.TestResult{Category: string(conformance.CatHeartbeat)}
				result, err := heartbeatDropRPC(ctx, 5)
				if err != nil {
					return tr, err
				}
				if result["configured"] != true {
					return tr, fmt.Errorf("expected configured true, got %v", result["configured"])
				}
				dropCount, ok := result["dropCount"].(float64)
				if !ok || dropCount != 5 {
					return tr, fmt.Errorf("expected dropCount 5, got %v", result["dropCount"])
				}
				tr.Passed = true
				return tr, nil
			},
		},
		{
			ID:       "HB-004",
			Category: conformance.CatHeartbeat,
			Name:     "pause heartbeat indefinite with pauseMs 0 returns duration indefinite",
			Run: func(ctx context.Context, host *conformance.PseudoHost) (conformance.TestResult, error) {
				tr := conformance.TestResult{Category: string(conformance.CatHeartbeat)}
				result, err := heartbeatPauseRPC(ctx, 0)
				if err != nil {
					return tr, err
				}
				if result["paused"] != true {
					return tr, fmt.Errorf("expected paused true, got %v", result["paused"])
				}
				duration, ok := result["duration"].(string)
				if !ok || duration != "indefinite" {
					return tr, fmt.Errorf("expected duration indefinite, got %v", result["duration"])
				}
				tr.Passed = true
				return tr, nil
			},
		},
		{
			ID:       "HB-005",
			Category: conformance.CatHeartbeat,
			Name:     "drop heartbeat with count 0 returns configured true and dropCount 0",
			Run: func(ctx context.Context, host *conformance.PseudoHost) (conformance.TestResult, error) {
				tr := conformance.TestResult{Category: string(conformance.CatHeartbeat)}
				result, err := heartbeatDropRPC(ctx, 0)
				if err != nil {
					return tr, err
				}
				if result["configured"] != true {
					return tr, fmt.Errorf("expected configured true, got %v", result["configured"])
				}
				dropCount, ok := result["dropCount"].(float64)
				if !ok || dropCount != 0 {
					return tr, fmt.Errorf("expected dropCount 0, got %v", result["dropCount"])
				}
				tr.Passed = true
				return tr, nil
			},
		},
		{
			ID:       "HB-006",
			Category: conformance.CatHeartbeat,
			Name:     "pause then resume then pause again works correctly",
			Run: func(ctx context.Context, host *conformance.PseudoHost) (conformance.TestResult, error) {
				tr := conformance.TestResult{Category: string(conformance.CatHeartbeat)}
				binary := conformance.FindPluginBinary()
				if binary == "" {
					return tr, fmt.Errorf("plugin binary not found")
				}
				h := conformance.NewPseudoHostForBinary(binary)
				if err := h.Start(ctx); err != nil {
					return tr, fmt.Errorf("start host: %w", err)
				}
				defer h.Kill()

				pausePayload := conformance.MustMarshal(map[string]any{"pauseMs": 1000})
				pauseResp, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.pause", pausePayload)
				if err != nil {
					return tr, fmt.Errorf("first pause rpc: %w", err)
				}
				var pauseResult map[string]any
				if err := json.Unmarshal(pauseResp.Payload, &pauseResult); err != nil {
					return tr, fmt.Errorf("unmarshal pause response: %w", err)
				}
				if pauseResult["paused"] != true {
					return tr, fmt.Errorf("expected first pause to return paused true")
				}

				resumeResp, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.resume", conformance.MustMarshal(map[string]any{}))
				if err != nil {
					return tr, fmt.Errorf("resume rpc: %w", err)
				}
				var resumeResult map[string]any
				if err := json.Unmarshal(resumeResp.Payload, &resumeResult); err != nil {
					return tr, fmt.Errorf("unmarshal resume response: %w", err)
				}
				if resumeResult["resumed"] != true {
					return tr, fmt.Errorf("expected resumed true")
				}

				pauseResp2, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.pause", conformance.MustMarshal(map[string]any{"pauseMs": 2000}))
				if err != nil {
					return tr, fmt.Errorf("second pause rpc: %w", err)
				}
				var pauseResult2 map[string]any
				if err := json.Unmarshal(pauseResp2.Payload, &pauseResult2); err != nil {
					return tr, fmt.Errorf("unmarshal second pause response: %w", err)
				}
				if pauseResult2["paused"] != true {
					return tr, fmt.Errorf("expected second pause to return paused true")
				}
				pauseMs, ok := pauseResult2["pauseMs"].(float64)
				if !ok || pauseMs != 2000 {
					return tr, fmt.Errorf("expected second pauseMs 2000, got %v", pauseResult2["pauseMs"])
				}
				tr.Passed = true
				return tr, nil
			},
		},
	}

	conformance.RunMatrix(t, conformance.CatHeartbeat, cases)
}

func heartbeatPauseRPC(ctx context.Context, pauseMs int) (map[string]any, error) {
	binary := conformance.FindPluginBinary()
	if binary == "" {
		return nil, fmt.Errorf("plugin binary not found")
	}

	h := conformance.NewPseudoHostForBinary(binary)
	if err := h.Start(ctx); err != nil {
		return nil, fmt.Errorf("start host: %w", err)
	}
	defer h.Kill()

	payload := conformance.MustMarshal(map[string]any{"pauseMs": pauseMs})
	resp, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.pause", payload)
	if err != nil {
		return nil, fmt.Errorf("heartbeat.pause rpc: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp.Payload, &result); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}

	return result, nil
}

func heartbeatDropRPC(ctx context.Context, count int) (map[string]any, error) {
	binary := conformance.FindPluginBinary()
	if binary == "" {
		return nil, fmt.Errorf("plugin binary not found")
	}

	h := conformance.NewPseudoHostForBinary(binary)
	if err := h.Start(ctx); err != nil {
		return nil, fmt.Errorf("start host: %w", err)
	}
	defer h.Kill()

	payload := conformance.MustMarshal(map[string]any{"count": count})
	resp, err := h.CallRPC(ctx, "mock.fault", "mock.fault.heartbeat.drop", payload)
	if err != nil {
		return nil, fmt.Errorf("heartbeat.drop rpc: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp.Payload, &result); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}

	return result, nil
}

File: testplugins/game-plugin-demo/conformance/ts/src/matrices/heartbeat.test.ts
import * as path from 'path';
import { PseudoHost } from '../harness';

const PLUGIN_PATH = path.resolve(__dirname, '../../../../typescript/dist/index.js');

describe('Heartbeat', () => {
  let host: PseudoHost;

  afterEach(() => {
    if (host) {
      host.kill();
    }
  });

  it('HB-001: pause heartbeat with pauseMs 500 returns paused true', async () => {
    host = new PseudoHost(PLUGIN_PATH);
    await host.start();
    const resp = await host.callRPC('mock.fault', 'mock.fault.heartbeat.pause', { pauseMs: 500 }, 5000);
    expect(resp.ok).toBe(true);
    const p = resp.payload as Record<string, unknown>;
    expect(p.paused).toBe(true);
    expect(p.pauseMs).toBe(500);
  });

  it('HB-002: resume heartbeat after pause returns paused false', async () => {
    host = new PseudoHost(PLUGIN_PATH);
    await host.start();
    const pauseResp = await host.callRPC('mock.fault', 'mock.fault.heartbeat.pause', { pauseMs: 2000 }, 5000);
    expect(pauseResp.ok).toBe(true);
    expect((pauseResp.payload as Record<string, unknown>).paused).toBe(true);

    const resumeResp = await host.callRPC('mock.fault', 'mock.fault.heartbeat.resume', {}, 5000);
    expect(resumeResp.ok).toBe(true);
    const p = resumeResp.payload as Record<string, unknown>;
    expect(p.paused).toBe(false);
  });

  it('HB-003: drop heartbeat with count 5 returns dropped count and remaining', async () => {
    host = new PseudoHost(PLUGIN_PATH);
    await host.start();
    const resp = await host.callRPC('mock.fault', 'mock.fault.heartbeat.drop', { count: 5 }, 5000);
    expect(resp.ok).toBe(true);
    const p = resp.payload as Record<string, unknown>;
    expect(p.dropped).toBe(5);
    expect(p.remaining).toBe(5);
  });

  it('HB-004: pause heartbeat with pauseMs 0 returns paused true indefinite', async () => {
    host = new PseudoHost(PLUGIN_PATH);
    await host.start();
    const resp = await host.callRPC('mock.fault', 'mock.fault.heartbeat.pause', { pauseMs: 0 }, 5000);
    expect(resp.ok).toBe(true);
    const p = resp.payload as Record<string, unknown>;
    expect(p.paused).toBe(true);
    expect(p.pauseMs).toBe(0);
  });

  it('HB-005: drop heartbeat counts accumulate across multiple calls', async () => {
    host = new PseudoHost(PLUGIN_PATH);
    await host.start();
    const first = await host.callRPC('mock.fault', 'mock.fault.heartbeat.drop', { count: 3 }, 5000);
    expect(first.ok).toBe(true);
    expect((first.payload as Record<string, unknown>).remaining).toBe(3);

    const second = await host.callRPC('mock.fault', 'mock.fault.heartbeat.drop', { count: 4 }, 5000);
    expect(second.ok).toBe(true);
    expect((second.payload as Record<string, unknown>).remaining).toBe(7);
  });

  it('HB-006: pause then resume then pause again works correctly', async () => {
    host = new PseudoHost(PLUGIN_PATH);
    await host.start();
    const firstPause = await host.callRPC('mock.fault', 'mock.fault.heartbeat.pause', { pauseMs: 1000 }, 5000);
    expect(firstPause.ok).toBe(true);
    expect((firstPause.payload as Record<string, unknown>).paused).toBe(true);

    const resume = await host.callRPC('mock.fault', 'mock.fault.heartbeat.resume', {}, 5000);
    expect(resume.ok).toBe(true);
    expect((resume.payload as Record<string, unknown>).paused).toBe(false);

    const secondPause = await host.callRPC('mock.fault', 'mock.fault.heartbeat.pause', { pauseMs: 2000 }, 5000);
    expect(secondPause.ok).toBe(true);
    const p = secondPause.payload as Record<string, unknown>;
    expect(p.paused).toBe(true);
    expect(p.pauseMs).toBe(2000);
  });
});
